MSU-USDA Python Workshop
# Python for Data Science: Getting Started with Python and Pandas

# I. Importing Necessary Packages
The following code will include the packages you'll need for this notebook. Packages are collections of code that adds additional functionality to the core Python.

Note: If you're new to Jupyter notebooks, to run the code, just click on the cell below and then click the "Run" button above in the menu. You can also use the shortcut of holding down "shift" and hitting the return key.

In [1]:
# Numpy supports multidimensional arrays and has a lot of math functions.
import numpy as np 

# Pandas is great for tabular data (data in rows and columns)
import pandas as pd

# Matplot Lib is for drawing plots, graphs, etc. and Seaborn is another package
# that makes matplotlib easier to use and create more attractive graphics
import matplotlib.pyplot as plt 
import matplotlib.image as mpimg 
import seaborn as sns 

# Geopandas adds geographical analysis capabilities to Pandas - one or more of your
# columns can be geographical (points, lines, shapes, etc.)
import geopandas as gpd
from geopandas import GeoDataFrame 

# These are built into core python but you have to include them to use them
import random 
import os 
import json 

# This package adds the ability to work with Excel files
import openpyxl

# II. Hello World
It's traditional for your first code in any language to be a "hello world." So, let's store "Hello World" in a variable and then print it out with python. 

This is a good time to turn on your line numbers through View -> Show Line Numbers.

In [4]:
myMessage = "Hello World, I'm Jon"
print(myMessage)

varName = "I love Milkshakes"
print(varName)

Hello World, I'm Jon
I love Milkshakes


# III. Python Variables and Types
In the previous section we stored the phrase "Hello World" in a variable named "myMessage." A variable is a container for data. Variables have *names* and they can contain data of many different *types* (numbers, character strings, etc.)

## What Types Does Python Support?
Python supports the following types by default:

| Kind of Data | Python Types | Examples |
| :----------- | :----------- | :----------- |
|Text|str|varName = "I love milkshakes."|
|Numeric|int, float, complex|varName = 3<br />varName = 3.0<br />varName = 3j|
|Sequence|list, tuple, range|varName = ["chocolate", "strawberry", "vanilla"]<br />varName = ("chocolate", "strawberry", "vanilla")<br />varName = range(6)
|Mapping|dict|varName = {"name" : "Bully", "species" : "Dog"}|
|Set|set, frozenset|varName = {"strawberry", "vanilla", "chocolate"}<br />varName = frozenset({"strawberry", "vanilla", "chocolate"})|
|Boolean|bool|varName = True|
|Binary|bytes, bytearray, memoryview|varName = b"hello"<br />varName=bytearray(5)<br />varName=memoryview(bytes(5))|
|None|NoneType|varName=None|

## Types That Don't Require Explanation
If you've programmed in another language, you will probably know what the following types are without any explanation:
- Text - str (text goes in what programmers call 'strings')
- Numeric - int (integers are counting numbers)
- Numeric - float (floating point numbers have a decimal and some amount of precision)
- Boolean - bool (boolean values are either True/1 or False/0)

## Types That Require Explanation
But there are some types that look a little weird at first, so let's walk through them.

### Numeric - complex
The *complex* type is used to store complex numbers. A complex number take the form a + ib, where a is the real number and ib is the imaginary number. i is the square root of negative one, so we call it "imaginary." One detail - engineers use the convention "j" instead of "i" for the square root of negative one, so Python follows the engineering convention. We won't be using these in this lab, but if you want to brush up on complex numbers, there's a [great intro at Khan Academy](https://www.khanacademy.org/math/algebra2/x2ec2f6f830c9fb89:complex/x2ec2f6f830c9fb89:imaginary/a/intro-to-the-imaginary-numbers).

### Sequences - list [] and tuple ()
Lists and tuples (pronounced too-pulls) are used to store lists of things like "1,2,3,4" or "dog,cat,horse,zebra". The difference is that lists can be changed after you create them (they are *mutable*) and tuples cannot be changed (they are *immutable*). The code below initializes a list and a tuple and illustrates how Python also lets you figure out what type any variable is.

In [7]:
# initialize a list with square brackets
myList = [8,7,19,9]

# initialize a tuple with parentheses
myTuple = (8,7,19,9)

# print out the type of each
print(type(myList))
print(type(myTuple))

#let's try to change a list; you should have no problem
myList[0] = 9

# print out that changed list - you'll see a 9 in the first position since you changed it
print(myList)

# you can also print out just the first member of the tuple or list like this
print(myTuple[0])

# but, let's try to change a tuple -  you should see an error
myTuple[0] = 9

# and so because of the error, this line will not execute at all
print(myTuple)

<class 'list'>
<class 'tuple'>
[9, 7, 19, 9]
8


TypeError: 'tuple' object does not support item assignment

Hey, don't worry about that error - I *knew* that would happen. Notice a few things about that code:
- Lists are assigned using square brackets [] and tuples are created using parentheses ()
- The type() method returns what type of variable something is (Lines 7 and 8)
- I can access an individual member of a list using brackets myList[0] and the number of the item (line 11). In the example above, I assign the first member of the list ([0] - all cool programming languages start counting from zero and not one) a new value and it switches from 8 to 9 when I print it out on line 14. You can access the second member of the list using myList[1], the third with myList[2], and so on. 
- Tuples are not changeable. Even though I can access the first item using [0] the same way I did with lists (line 17), when I try to change myTuple, I get a "TypeError" - because you can can't change tuples once they are created.

So, line 20 generates an error because I tried to change a tuple. Tuples exist because they are much faster to process than lists. If you have a list of things that won't change while a program runs, storing it as a tuple means faster code.

### Range (start,stop,step)
Range is used to hold a series of numbers. Suppose you are going to do some research on the multi-state lottery in which numbers range from 10 to 45. You could declare that as a range and then use that range throughout your code for all kinds of purposes. 

range(start,stop,step) is exclusive, like in the example given range(10,46,1) 46 wouldn't be included. 

For example:

In [12]:
# lottery ticket balls range from 10 to 45
# let's assign this range to the variable "lotteryRange"
lotteryRange = range(10,46,1)  # range(start,stop,step)

# loop through the range and print out every number
# notice that at each iteration of the loop, the variable "number" will contain the next
# number in the range
for number in lotteryRange:
    print(number)

10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
Jon


As you can see, our range goes from 10 to 45 incremented by 1. If you leave out the start number and the step size, Python assumes you want to start at 0 and increment by 1. In Python, a range is *immutable*. This example also used a "for" loop meaning that the block of code below it executes many times. In this case the number of times is determined by how many members there are in the range of numbers.

### Sidebar: Indentation in Python

Notice that after the "for" loop starts there is a colon and then the lines contained within the loop are indented by four spaces. Python doesn't use curly brackets like C-like languages do, but uses indentation to mark out blocks of code. Jupyter will take care of this for you for the most part. If you are using a plain text editor, just be consistent, you have to use a consistent number of spaces for the indent or a tab, but you can't mix styles.

### Using Range in a For Loop

Range is often used in the context of a *for loop* like this:

In [13]:
# range here is called with "from" and "to" and the default increment of 1 will be
# in effect
for counter in range(0,10):
    print(counter)

0
1
2
3
4
5
6
7
8
9


### Mapping - Dictionary {"key":"value","key2":"value2"}

Dictionaries are very useful for data science. They consist of a list of "key / value" pairs. Just like a dictionary has a word (key) and a definition (value), Python dictionaries have these pairs of keys/values. For example, here's some code that stores US Department of Labor occupation codes in a dictionary:


In [17]:
#lets initialize our dictionary to contain three occupations
# notice that dictionaries use curly brackets
occupationCodes = {"welder" : "51-4121.00", "nurse" : "29-1141.00", "computer programmer" : "15-1251.00"}

print(occupationCodes["welder"])

#oops, I forgot one, I can also add one at at a time
occupationCodes["accountant"] = "13-2011.00"

# a customer typed in a keyword to search for a key in the dictionary
# note: if this were real code you would need to "sanitize" your inputs
# and make sure the customer didn't type anything that could potentially
# ruin your database or expose a system vulnerability
keyword = "welder"

#now, I can do things with the dictionary like check if there is an entry:
if keyword in occupationCodes:
    print(f"The occupation code for {keyword} is {occupationCodes.get(keyword)}")
else:
    print(f"Sorry, don't have information on {keyword}")


51-4121.00
The occupation code for welder is 51-4121.00


There are lots of cool things to notice in this code:
- You can modify dictionaries (line 6)
- You can check to see whether a key exists in the dictionary - on line 15 we check to see if the customer's keyword shows up as a key in the dictionary.
- You can use .get(key) with a dictionary to retrieve the corresponding value (line 16)
- You can not only print variables using the {} curly brackets, but you can also put code to execute in the curly bracket (line 16).

Try modifying the keyword to something like "firefighter" that we know isn't in the dictionary and then running it.

This code snippet above also uses the "if ... else" functionality in Python. The next section after we finish with types includes an explanation of this way to *check conditions* before executing code.

Dictionaries, starting in Python 3.7 are *ordered* collections which means you can count on key/value pairs being stored in the order in which you added them to the dictionary.

As awesome as dictionaries are, we will also use "data frames" later in this class to store multi-dimensional data. This functionality won't come from the basic Python language itself, it will be drawn from a library called "Pandas."

### Set {} and FrozenSet (())
While *list* and *tuple* have an order to them and you can access their values by referring to the order using brackets ([0] grabs the first item), *set* and *frozen set* are not ordered and you can't reference individual members using an index in brackets. It's like a library full of books with no card catalog and a blindfolded librarian is only able to hand you a book at random (though not truly randomly). 

With *set*, you can remove items from or add items to the *set*, but you can't modify an item directly. With a *list*, you could change "Charles" to Charlie." But with a *set*, you'd have to delete "Charles" and then add "Charlie." You also can't have duplicate values in a *set*. So you could represent the gameplay of duck, duck, goose using a *list*, but not a *set*. A *set* would only store "duck" and "goose" once each. 

The code below illustrates the properties of a set:

In [18]:
# Short way to create a set (notice curly brackets)
universitySet = {"Mississippi State", "Michigan State", "Columbia", "LSU", "Tulane", "Notre Dame"}

print(universitySet)

print(f"This set contains {len(universitySet)} items.")

if "USM" in universitySet:
    print("Cool, USM is Represented")
else:
    print("We need to add USM")
    # add a college
    universitySet.add("USM")

# remove a college
universitySet.remove("Notre Dame")

print(universitySet) 

# Longer way to create a set (using the 'set' constructor method)
communityCollegeSet = set(("Jones College","EMCC","Meridian Community College"))
print(communityCollegeSet)

{'Tulane', 'Columbia', 'Notre Dame', 'LSU', 'Mississippi State', 'Michigan State'}
This set contains 6 items.
We need to add USM
{'Tulane', 'Columbia', 'LSU', 'Mississippi State', 'Michigan State', 'USM'}
{'Jones College', 'Meridian Community College', 'EMCC'}


Notice a few things:
- There are two ways to initialize a set - the shorthand (line 2) and the longhand (line 17)
- The method "len()" tells you how many members are in a set (line 6). You could also use len with a dictionary or other collections in python.
- We can use "in" to see if a value is contained in the set (line 8)
- You can .add() (line 13) and .remove() (line 16) items from the set

A frozenset is like a set, but you cannot add or remove values once it is created. This code creates a frozenset then generates an error when trying to change it:

In [19]:
communityCollegeSet = frozenset(("Jones College","EMCC","Meridian Community College"))
print(communityCollegeSet)

communityCollegeSet.remove("EMCC")
print(communityCollegeSet)

frozenset({'Jones College', 'Meridian Community College', 'EMCC'})


AttributeError: 'frozenset' object has no attribute 'remove'

You can't remove "EMCC" from the list because frozensets are immutable and Python enforces this.

Sets use a "hash table" under the hood so they are super fast for figuring out whether something is in the set. That's their main advantage: speed.

## Figuring Out What a Variable Can Do

A cool trick is that you can type a variable name in Python into a Jupyter notebook, then a period after it, then hit the "tab" and you'll see a list of possible functions that are available for that instance. Give it a shot - just put your cursor after the variable below and you'll see a list of the functions available.

In [ ]:
communityCollegeSet.

It should look like this:

<img src="https://dsci.msstate.edu/downloads/USDA/autocomplete.png" style="width:600px" />    

# IV. Iterating through Lists and Dictionaries
One thing you'll need to do quite frequently in data science is to iterate through the values in a list, one by one. There are a few ways to do this. Here's the first:

In [24]:
# Create a List
greatLanguages = ["Python", "R", "Rust", "C++"]
for language in greatLanguages:
    print(f"{language} is a terrible language.")

Python is a terrible language.
R is a terrible language.
Rust is a terrible language.
C++ is a terrible language.


Notice that:
- The variable "language" stores, one by one, the words in the greatLanguages collection (Line 2)
- Within the for loop, at each step, you can operate against the variable language and do things like printing (Line 4)

Sometimes you want to iterate in such a way as to store the index of each item and not just get the value of each item in the list:

In [26]:
greatLanguages = ["Python", "R", "Rust", "C++"]
for position, language in enumerate(greatLanguages):
    print(position, language)

0 Python
1 R
2 Rust
3 C++


In the example above:
- The variable *position* contains the index number of the item at each iteration (0-3)
- The variable *language* contains the value of the item (just as before)
- The "enumerate" method is what extracts the value of the *position* and the *language variables*

Iterating through a dictionary works similarly:

In [29]:
#lets initialize our dictionary to contain three occupation titles with their corresponding codes
occupationCodes = {"welder" : "51-4121.00", "nurse" : "29-1141.00", "computer programmer" : "15-1251.00"}

for key, value in occupationCodes.items():
    print(key, " is coded as ", value)


welder  is coded as  51-4121.00
nurse  is coded as  29-1141.00
computer programmer  is coded as  15-1251.00


### Exercise 1: List

Create a list of your favorite band names and then iterate through the list and output each one in the format like "I love The Rolling Stones".

In [31]:
faveBands = ["The Rolling Stones", "REM", "The Pixies"]
for band in faveBands:
    print(f"I love {band}")

I love The Rolling Stones
I love REM
I love The Pixies


# V. Functions

You're familiar now with Python's built-in functions like "print." Sometimes you'll need to create custom functions.

For example, let's create and use a trivial function that takes three values and returns their average.

In [32]:
# define our function
def averageThreeVars(var1,var2,var3):
    # calculate the average of the three
    average = (var1 + var2 + var3) / 3
    # return the average
    return(average)

# let's find the average height of the three highest mountains in the world

# store the height in feet of three mountains in separate variables
everestHeight = 29032
k2Height = 28251
kangchenjungaHeight = 28169

# call the function averageThreeVars and pass in our three peaks, store the
# result in a global variable called averageHeightThreePeaks
averageHeightThreePeaks = averageThreeVars(everestHeight,k2Height,kangchenjungaHeight)

# print out our global variable
print(f"The average height of our three peaks is {averageHeightThreePeaks}")


The average height of our three peaks is 28484.0


Notice that our function received the three mountain heights, calculated an average, and then "returned" that average where we stored it into a global variable called "averageHeightThreePeaks."

In most languages, you can only return one variable from a function. That variable can be a collection (e.g., a list) and so this isn't very limiting, but Python is pretty unique in being able to return multiple values from one function. We won't go into that today, but keep that in mind for the future.

## Lambda Functions

So far, this notebook has focused on defined functions. But there is another kind of function in Python called a "Lambda" function. This is probably the most advanced topic we cover here, but you'll see us use it later and may want to come back to review this.

The following sample code will call a traditional function and a lambda version of the same function to add three to the number we give the function.

In [33]:
# we have some variable we want to change
myVariable = 7

# define a traditional function
def addThreeTraditional(number):
    return number+3

# call it and pass in myVariable to add 3
myResult = addThreeTraditional(myVariable)

# print the result
print(myResult)

# now let's see a lambda version
addThreeLambda = lambda x: x + 3

# call the lambda
myResult = addThreeLambda(myVariable)

# print the result
print(myResult)

10
10


Note that you should get the same result for both the regular and lambda versions of the function (10).

Why on earth would we use lambdas? Well, a key reason would be that we don't have to define them ahead of time like we did above. The following is equivalent but it uses what are called "anonymous" functions - a lambda with no name.


In [34]:
# get a result by, instead of using "addThreeLambda", just put the lambda 
# in parentheses and call with an argument in parentheses

myResult = (lambda x:x+3)(myVariable)
print(myResult)


10


Amazing, right? You can also pass in multiple variables just like with a defined function.

In [35]:
var1 = "Amazing"
var2 = "Spiderman"

# notice our lambda has two variables - x and y
# and so we pass in the arguments in order
superhero = (lambda x,y:x + " " + y)(var1,var2)

print(superhero)

Amazing Spiderman


So we passed in two variables to the lambda, it squished them together with a space between and then returned them. You should see "Amazing Spiderman."

Now, just so you know the terminology of the parts of a lambda:

<img src="https://dsci.msstate.edu/downloads/wrangling/lab6/lambda.png" alt="parts of a lambda" width="400"/>


## Python: Learning More

Well, there's a lot more to Python, but the above gives you a start on variables, loops, collections, and iterating through collections.

The great thing about Python is that you can google "How do I (do whatever) in python?" and you'll get help in many places.

# VI. Tabular Data with PANDAS

The [pandas website](https://pandas.pydata.org/) describes the project's goal - pandas "aims to be the fundamental high-level building block for doing practical, real world data analysis in Python. Additionally, it has the broader goal of becoming the most powerful and flexible open source data analysis / manipulation tool available in any language."

PANDAS uses two kinds of structures to store data - *data frames* and *series*. Data Frames are like an excel spreadsheet - think of a table of data with headings and values. Series are like a single column in an excel spreadsheet. Both types of data have an *index* - you can think of an index, for now, as a line number, but it can be anything, even text.

<img src="https://pandas.pydata.org/docs/_images/01_table_dataframe.svg">


(Image Source: Pandas Tutorial, https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html)

## Making a Data Frame from Scratch

You can make a data frame programmatically. For example, let's create a data frame from lists.

In [36]:
listOfTuples = [(1,'089',32223,59730,98468,35297),
                   (2,'121',27183,56159,145165,52539),
                   (3,'073',27399,50075,57786,21237),
                   (4,'033',25065,59734,166234,56641),
                   (5,'059',23547,49620,140298,50185),
                   (6,'047',23111,44550,194029,69384),
                   (7,'149',22079,40404,48773,18941),
                   (8,'045',21935,44494,43929,17380),
                   (9,'081',21831,39049,82910,32086),
                   (10,'131',21691,43728,17786,6165)]

listOfColumnNames = ["rank","county_fips","per_capita_income"
                     ,"median_household_income","population"
                     ,"num_households"]

countyData = pd.DataFrame(listOfTuples,columns=listOfColumnNames)
        
# if placed on a line by itself, you will get pretty output of the data frame    
countyData

rank county_fips  per_capita_income  median_household_income  population  \
0     1         089              32223                    59730       98468   
1     2         121              27183                    56159      145165   
2     3         073              27399                    50075       57786   
3     4         033              25065                    59734      166234   
4     5         059              23547                    49620      140298   
5     6         047              23111                    44550      194029   
6     7         149              22079                    40404       48773   
7     8         045              21935                    44494       43929   
8     9         081              21831                    39049       82910   
9    10         131              21691                    43728       17786   

   num_households  
0           35297  
1           52539  
2           21237  
3           56641  
4           50185  
5           69384  
6           18941  
7           17380  
8           32086  
9            6165

## CSV Files

Pandas makes it very easy to load an excel file or other tabular data sources into data frames. 

Generally, we use .csv files instead of excel for common data exchanges because excel has a limit on length (1,048,576 rows). CSV is a common data format that uses a comma between the value in each column of data ("comma delimited") and it usually puts quotation marks around any data that inherently has commas (in case one of your columns itself needs commas).

For example, a raw CSV file with a header row might look like this:

<pre>
Book Title,Publisher,Price
War and Peace,Vintage Classics,12.99
"Our Bodies, Ourselves",Touchstone,48.38
Putin's Playbook,"Simon & Schuster, Inc.",14.49
</pre>

Notice that the second book listed has a comma in the title, so it is surrounded by quotation marks to avoid our interpreting the comma as a new column.

The third book listed has a comma in the publisher name, so the quotation marks are used.

## Loading a .csv File from a URL

Here's an example of using pandas to load college football bowl data. It will load a file that is .csv format (comma-separated-values) into data frames. To make it easy at first we'll just load a file from the internet so you don't have to worry about local paths:

In [39]:
# load in a csv file (this one is remotely located on a server)
# note - the data dictionary is in https://dsci.msstate.edu/downloads/wrangling/lab2/collegefootballbowl.txt

bowlData = pd.read_csv ('https://dsci.msstate.edu/downloads/wrangling/lab2/collegefootballbowl.csv')

bowlData

id  year        date  day         winner_tie winner_rank  \
0        1  2021  12/29/2021  Wed           Oklahoma          14   
1        2  2020  12/29/2020  Tue              Texas          20   
2        3  2019  12/31/2019  Tue              Texas         NaN   
3        4  2018  12/28/2018  Fri   Washington State          12   
4        5  2017  12/28/2017  Thu    Texas Christian          13   
...    ...   ...         ...  ...                ...         ...   
1522  1523  2004  12/30/2004  Thu  Northern Illinois         NaN   
1523  1524  2003  12/30/2003  Tue       Fresno State         NaN   
1524  1525  2002  12/31/2002  Tue       Fresno State         NaN   
1525  1526  2001  12/31/2001  Mon     Michigan State         NaN   
1526  1527  2000  12/31/2000  Sun          Air Force         NaN   

      winner_points     loser_tie loser_rank  loser_points  attendance  \
0                47        Oregon         15            32     59121.0   
1                55      Colorado        NaN            23     10822.0   
2                38          Utah         12            10     60147.0   
3                28    Iowa State         25            26     60675.0   
4                39      Stanford         15            37     57653.0   
...             ...           ...        ...           ...         ...   
1522             34          Troy        NaN            21     21456.0   
1523             17          UCLA        NaN             9     20126.0   
1524             30  Georgia Tech        NaN            21     10132.0   
1525             44  Fresno State         20            35     30456.0   
1526             37  Fresno State        NaN            34     26542.0   

                                                    mvp sponsor  \
0     Oklahoma RB Kennedy Brooks, Oklahoma S Pat Fields  Valero   
1     Texas RB Bijan Robinson, Texas LB DeMarvion Ov...  Valero   
2          Texas QB Sam Ehlinger, Texas LB Joseph Ossai  Valero   
3     Washington State QB Gardner Minshew, Washingto...  Valero   
4               TCU QB Kenny Hill, TCU LB Travin Howard  Valero   
...                                                 ...     ...   
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...     NaN   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...     NaN   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...     NaN   
1525  WR Charles Rogers (Michigan State), DL Nick My...     NaN   
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...     NaN   

                bowl_name  
0              Alamo Bowl  
1              Alamo Bowl  
2              Alamo Bowl  
3              Alamo Bowl  
4              Alamo Bowl  
...                   ...  
1522  Silicon Valley Bowl  
1523  Silicon Valley Bowl  
1524  Silicon Valley Bowl  
1525  Silicon Valley Bowl  
1526  Silicon Valley Bowl  

[1527 rows x 14 columns]

### Head(): Viewing Just the first N Lines
Use head() to get a quick look at the first part of the data file.

In [40]:
## Head and Tail

bowlData.head(10)

id  year        date  day        winner_tie winner_rank  winner_points  \
0   1  2021  12/29/2021  Wed          Oklahoma          14             47   
1   2  2020  12/29/2020  Tue             Texas          20             55   
2   3  2019  12/31/2019  Tue             Texas         NaN             38   
3   4  2018  12/28/2018  Fri  Washington State          12             28   
4   5  2017  12/28/2017  Thu   Texas Christian          13             39   
5   6  2016  12/29/2016  Thu    Oklahoma State          13             38   
6   7  2015    1/2/2016  Sat   Texas Christian          11             47   
7   8  2014    1/2/2015  Fri              UCLA          14             40   
8   9  2013  12/30/2013  Mon            Oregon          10             30   
9  10  2012  12/29/2012  Sat             Texas         NaN             31   

      loser_tie loser_rank  loser_points  attendance  \
0        Oregon         15            32     59121.0   
1      Colorado        NaN            23     10822.0   
2          Utah         12            10     60147.0   
3    Iowa State         25            26     60675.0   
4      Stanford         15            37     57653.0   
5      Colorado         11             8     59815.0   
6        Oregon         15            41     64569.0   
7  Kansas State         11            35     60517.0   
8         Texas        NaN             7     65918.0   
9  Oregon State         15            27     65277.0   

                                                 mvp  \
0  Oklahoma RB Kennedy Brooks, Oklahoma S Pat Fields   
1  Texas RB Bijan Robinson, Texas LB DeMarvion Ov...   
2       Texas QB Sam Ehlinger, Texas LB Joseph Ossai   
3  Washington State QB Gardner Minshew, Washingto...   
4            TCU QB Kenny Hill, TCU LB Travin Howard   
5     OSU WR James Washington, OSU DT Vincent Taylor   
6       TCU QB Brian Kohlhausen, TCU S Travin Howard   
7   RB Paul Perkins (UCLA), LB Eric Kendricks (UCLA)   
8  QB Marcus Mariota (Oregon), SS Avery Patterson...   
9  WR Marquise Goodwin (Texas), DL Alex Okafor (T...   

                     sponsor   bowl_name  
0                     Valero  Alamo Bowl  
1                     Valero  Alamo Bowl  
2                     Valero  Alamo Bowl  
3                     Valero  Alamo Bowl  
4                     Valero  Alamo Bowl  
5                     Valero  Alamo Bowl  
6                     Valero  Alamo Bowl  
7                     Valero  Alamo Bowl  
8  Valero Energy Corporation  Alamo Bowl  
9  Valero Energy Corporation  Alamo Bowl

### Tail(): Viewing Just the last N Lines
Use tail() to get a quick look at the end of the data file.

In [41]:
bowlData.tail(10)

id  year        date  day         winner_tie winner_rank  \
1517  1518  2001  12/27/2001  Thu       Georgia Tech         NaN   
1518  1519  2000  12/24/2000  Sun            Georgia          24   
1519  1520  1999  12/25/1999  Sat             Hawaii         NaN   
1520  1521  1998  12/25/1998  Fri          Air Force          16   
1521  1522  1948  12/18/1948  Sat     Hardin-Simmons         NaN   
1522  1523  2004  12/30/2004  Thu  Northern Illinois         NaN   
1523  1524  2003  12/30/2003  Tue       Fresno State         NaN   
1524  1525  2002  12/31/2002  Tue       Fresno State         NaN   
1525  1526  2001  12/31/2001  Mon     Michigan State         NaN   
1526  1527  2000  12/31/2000  Sun          Air Force         NaN   

      winner_points     loser_tie loser_rank  loser_points  attendance  \
1517             24      Stanford         11            14     30144.0   
1518             37      Virginia        NaN            14     24187.0   
1519             23  Oregon State        NaN            17     40974.0   
1520             45    Washington        NaN            25     46451.0   
1521             40      Ouachita        NaN            12         NaN   
1522             34          Troy        NaN            21     21456.0   
1523             17          UCLA        NaN             9     20126.0   
1524             30  Georgia Tech        NaN            21     10132.0   
1525             44  Fresno State         20            35     30456.0   
1526             37  Fresno State        NaN            34     26542.0   

                                                    mvp sponsor  \
1517                                                NaN    Jeep   
1518                                                NaN    Jeep   
1519                                                NaN    Jeep   
1520                                                NaN    Jeep   
1521                                                NaN     NaN   
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...     NaN   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...     NaN   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...     NaN   
1525  WR Charles Rogers (Michigan State), DL Nick My...     NaN   
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...     NaN   

                bowl_name  
1517         Seattle Bowl  
1518         Seattle Bowl  
1519         Seattle Bowl  
1520         Seattle Bowl  
1521          Shrine Bowl  
1522  Silicon Valley Bowl  
1523  Silicon Valley Bowl  
1524  Silicon Valley Bowl  
1525  Silicon Valley Bowl  
1526  Silicon Valley Bowl

## Investigating the Data

### Shape Property: How Many Rows / Columns?

In [43]:
bowlData.shape

(1527, 14)

So we see there are 1527 rows and 14 columns.

### Dtypes Property: Understanding the Data Types of Each Column

Maybe you would like to know the data type of each column. Just run:

In [44]:
bowlData.dtypes

id                 int64
year               int64
date              object
day               object
winner_tie        object
winner_rank       object
winner_points      int64
loser_tie         object
loser_rank        object
loser_points       int64
attendance       float64
mvp               object
sponsor           object
bowl_name         object
dtype: object

Anything column that says "object" is probably storing strings. int64 and float64 are numbers.

### Describe(): Summary of Numerical Data (min, max, count, mean, quartiles)

Maybe you would like to have a summary of the data. Just run:

In [45]:
bowlData.describe()

id         year  winner_points  loser_points     attendance
count  1527.000000  1527.000000    1527.000000   1527.000000    1518.000000
mean    764.000000  1991.286182      30.253438     17.092338   49487.574440
std     440.951244    24.437900      12.111077     10.395141   23552.602532
min       1.000000  1901.000000       0.000000      0.000000       0.000000
25%     382.500000  1976.000000      21.000000     10.000000   31383.000000
50%     764.000000  1998.000000      30.000000     16.000000   49056.000000
75%    1145.500000  2011.000000      38.000000     24.000000   68321.500000
max    1527.000000  2021.000000      70.000000     61.000000  106869.000000

Notice that .describe returned a description of all the numerical columns including number of values, mean, standard deviation, min, max, and percentiles. The "min" of the year column is interesting - 1901 is the earliest year on record.

Also, notice that we only get descriptions of numerical columns. It's weird that we don't see the column for winner_rank, for example, but that's because it contains some non-numeric data. It's "messy" like a lot of real data is.

To see all columns, numeric or not, you can do this:

In [46]:
bowlData.describe(include = 'all')

id         year      date   day winner_tie winner_rank  \
count   1527.000000  1527.000000      1527  1527       1527         754   
unique          NaN          NaN       691     7        166          26   
top             NaN          NaN  1/1/1948   Sat    Alabama           3   
freq            NaN          NaN        11   412         45          49   
mean     764.000000  1991.286182       NaN   NaN        NaN         NaN   
std      440.951244    24.437900       NaN   NaN        NaN         NaN   
min        1.000000  1901.000000       NaN   NaN        NaN         NaN   
25%      382.500000  1976.000000       NaN   NaN        NaN         NaN   
50%      764.000000  1998.000000       NaN   NaN        NaN         NaN   
75%     1145.500000  2011.000000       NaN   NaN        NaN         NaN   
max     1527.000000  2021.000000       NaN   NaN        NaN         NaN   

        winner_points loser_tie loser_rank  loser_points     attendance  \
count     1527.000000      1527        671   1527.000000    1518.000000   
unique            NaN       171         26           NaN            NaN   
top               NaN   Alabama          2           NaN            NaN   
freq              NaN        30         38           NaN            NaN   
mean        30.253438       NaN        NaN     17.092338   49487.574440   
std         12.111077       NaN        NaN     10.395141   23552.602532   
min          0.000000       NaN        NaN      0.000000       0.000000   
25%         21.000000       NaN        NaN     10.000000   31383.000000   
50%         30.000000       NaN        NaN     16.000000   49056.000000   
75%         38.000000       NaN        NaN     24.000000   68321.500000   
max         70.000000       NaN        NaN     61.000000  106869.000000   

                                 mvp             sponsor  bowl_name  
count                           1358                 820       1527  
unique                          1331                 166         77  
top     QB Byron Leftwich (Marshall)  Outback Steakhouse  Rose Bowl  
freq                               3                  26        108  
mean                             NaN                 NaN        NaN  
std                              NaN                 NaN        NaN  
min                              NaN                 NaN        NaN  
25%                              NaN                 NaN        NaN  
50%                              NaN                 NaN        NaN  
75%                              NaN                 NaN        NaN  
max                              NaN                 NaN        NaN

You can now see "top" which gives you the most common value in the column.

### Analyzing Individual Columns

You can also run mathematical measures of central tendency against individual columns. For example:

In [48]:
# the mean of winner_points
print("Winners had an average of ",bowlData['winner_points'].mean())

# the median of loser_points
print("Losers had a median of ",bowlData['winner_points'].median())

Winners had an average of  30.25343811394892
Losers had a median of  30.0


### Info(): Understanding all Fields, Null Values, Size, etc.

In [49]:
# the .info method, like the property .dtypes, is great for figuring out the shape of our data
bowlData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1527 entries, 0 to 1526
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             1527 non-null   int64  
 1   year           1527 non-null   int64  
 2   date           1527 non-null   object 
 3   day            1527 non-null   object 
 4   winner_tie     1527 non-null   object 
 5   winner_rank    754 non-null    object 
 6   winner_points  1527 non-null   int64  
 7   loser_tie      1527 non-null   object 
 8   loser_rank     671 non-null    object 
 9   loser_points   1527 non-null   int64  
 10  attendance     1518 non-null   float64
 11  mvp            1358 non-null   object 
 12  sponsor        820 non-null    object 
 13  bowl_name      1527 non-null   object 
dtypes: float64(1), int64(4), object(9)
memory usage: 167.1+ KB


## Selecting Data

To find a specific set of rows, columns, or cells, you can use pandas to select data. There are several ways to do this.

### The df.loc[] approach:

With **.loc[rows, columns]** you can directly access rows and columns by name. If you have an integer in the "row" part, Pandas assumes this is the index of the row.

#### Single Cell

In [51]:
# SELECT a single cell - the attendance where row index is 1
bowlData.loc[1 , 'attendance']

10822.0

#### Single row by index

In [52]:
# SELECT whole row of data where row index is 1
bowlData.loc[1 , :]

id                                                               2
year                                                          2020
date                                                    12/29/2020
day                                                            Tue
winner_tie                                                   Texas
winner_rank                                                     20
winner_points                                                   55
loser_tie                                                 Colorado
loser_rank                                                     NaN
loser_points                                                    23
attendance                                                 10822.0
mvp              Texas RB Bijan Robinson, Texas LB DeMarvion Ov...
sponsor                                                     Valero
bowl_name                                               Alamo Bowl
Name: 1, dtype: object

#### Single Column by Column Name

In [53]:
# SELECT a single column of data, just the atttendance column
bowlData.loc[: , 'attendance']

0       59121.0
1       10822.0
2       60147.0
3       60675.0
4       57653.0
         ...   
1522    21456.0
1523    20126.0
1524    10132.0
1525    30456.0
1526    26542.0
Name: attendance, Length: 1527, dtype: float64

#### Slice of Rows by Index Range

In [54]:
# SELECT a "slice" of rows where index is between 1 and 20
bowlData.loc[1:20 , :]

id  year        date  day        winner_tie winner_rank  winner_points  \
1    2  2020  12/29/2020  Tue             Texas          20             55   
2    3  2019  12/31/2019  Tue             Texas         NaN             38   
3    4  2018  12/28/2018  Fri  Washington State          12             28   
4    5  2017  12/28/2017  Thu   Texas Christian          13             39   
5    6  2016  12/29/2016  Thu    Oklahoma State          13             38   
6    7  2015    1/2/2016  Sat   Texas Christian          11             47   
7    8  2014    1/2/2015  Fri              UCLA          14             40   
8    9  2013  12/30/2013  Mon            Oregon          10             30   
9   10  2012  12/29/2012  Sat             Texas         NaN             31   
10  11  2011  12/29/2011  Thu            Baylor          15             67   
11  12  2010  12/29/2010  Wed    Oklahoma State          16             36   
12  13  2009    1/2/2010  Sat        Texas Tech         NaN             41   
13  14  2008  12/29/2008  Mon          Missouri          25             30   
14  15  2007  12/29/2007  Sat        Penn State         NaN             24   
15  16  2006  12/30/2006  Sat             Texas          18             26   
16  17  2005  12/28/2005  Wed          Nebraska         NaN             32   
17  18  2004  12/29/2004  Wed        Ohio State          24             33   
18  19  2003  12/29/2003  Mon          Nebraska          22             17   
19  20  2002  12/28/2002  Sat         Wisconsin         NaN             31   
20  21  2001  12/29/2001  Sat              Iowa         NaN             19   

         loser_tie loser_rank  loser_points  attendance  \
1         Colorado        NaN            23     10822.0   
2             Utah         12            10     60147.0   
3       Iowa State         25            26     60675.0   
4         Stanford         15            37     57653.0   
5         Colorado         11             8     59815.0   
6           Oregon         15            41     64569.0   
7     Kansas State         11            35     60517.0   
8            Texas        NaN             7     65918.0   
9     Oregon State         15            27     65277.0   
10      Washington        NaN            56     65256.0   
11         Arizona        NaN            10     57593.0   
12  Michigan State        NaN            31     64757.0   
13    Northwestern         22            23     55986.0   
14       Texas A&M        NaN            17     66166.0   
15            Iowa        NaN            24     65875.0   
16        Michigan         20            28     62000.0   
17  Oklahoma State        NaN             7     65265.0   
18  Michigan State        NaN             3     56226.0   
19        Colorado         14            28     50690.0   
20      Texas Tech        NaN            16     65232.0   

                                                  mvp  \
1   Texas RB Bijan Robinson, Texas LB DeMarvion Ov...   
2        Texas QB Sam Ehlinger, Texas LB Joseph Ossai   
3   Washington State QB Gardner Minshew, Washingto...   
4             TCU QB Kenny Hill, TCU LB Travin Howard   
5      OSU WR James Washington, OSU DT Vincent Taylor   
6        TCU QB Brian Kohlhausen, TCU S Travin Howard   
7    RB Paul Perkins (UCLA), LB Eric Kendricks (UCLA)   
8   QB Marcus Mariota (Oregon), SS Avery Patterson...   
9   WR Marquise Goodwin (Texas), DL Alex Okafor (T...   
10  RB Terrance Ganaway (Baylor), LB Elliot Coffey...   
11  WR Justin Blackmon (Oklahoma State), DB Markel...   
12  QB Taylor Potts (Texas Tech), DB Jamar Wall (T...   
13  WR Jeremy Maclin (Missouri), LB Sean Witherspo...   
14  RB Rodney Kinlaw (Penn State), LB Sean Lee (Pe...   
15       QB Colt McCoy (Texas), DB Aaron Ross (Texas)   
16   RB Cory Ross (Nebraska), DB Leon Hall (Michigan)   
17  WR Ted Ginn Jr. (Ohio State), DL Simon Fraser ...   
18  QB Jammal Lord (Nebraska), DL Trevor Johnson (...   
19  QB Brooks Bollinger (Wisconsin), LB Je

#### Slice Containing All Rows, but Only Certain Columns

In [55]:
# SELECT all rows but only a  "slice" of columns 
bowlData.loc[: , 'year':'winner_rank']

year        date  day         winner_tie winner_rank
0     2021  12/29/2021  Wed           Oklahoma          14
1     2020  12/29/2020  Tue              Texas          20
2     2019  12/31/2019  Tue              Texas         NaN
3     2018  12/28/2018  Fri   Washington State          12
4     2017  12/28/2017  Thu    Texas Christian          13
...    ...         ...  ...                ...         ...
1522  2004  12/30/2004  Thu  Northern Illinois         NaN
1523  2003  12/30/2003  Tue       Fresno State         NaN
1524  2002  12/31/2002  Tue       Fresno State         NaN
1525  2001  12/31/2001  Mon     Michigan State         NaN
1526  2000  12/31/2000  Sun          Air Force         NaN

[1527 rows x 5 columns]

#### Slice Containing Certain Rows, Certain Columns

In [56]:
# SELECT a "slice" of cells where row index is between 10 and 20 and only a few columns
bowlData.loc[10:20 , 'year':'winner_rank']

year        date  day      winner_tie winner_rank
10  2011  12/29/2011  Thu          Baylor          15
11  2010  12/29/2010  Wed  Oklahoma State          16
12  2009    1/2/2010  Sat      Texas Tech         NaN
13  2008  12/29/2008  Mon        Missouri          25
14  2007  12/29/2007  Sat      Penn State         NaN
15  2006  12/30/2006  Sat           Texas          18
16  2005  12/28/2005  Wed        Nebraska         NaN
17  2004  12/29/2004  Wed      Ohio State          24
18  2003  12/29/2003  Mon        Nebraska          22
19  2002  12/28/2002  Sat       Wisconsin         NaN
20  2001  12/29/2001  Sat            Iowa         NaN

#### Rows Where Column has Certain Value

In [57]:
# this will find the locations in the dataset where year is equal to 1901
bowlData.loc[bowlData['year'] == 1901]

id  year      date  day winner_tie winner_rank  winner_points  \
1020  1021  1901  1/1/1902  Wed   Michigan         NaN             49   

     loser_tie loser_rank  loser_points  attendance                      mvp  \
1020  Stanford        NaN             0      8000.0  RB Neil Snow (Michigan)   

     sponsor  bowl_name  
1020     NaN  Rose Bowl

This returned only one row. You can read all about that season of bowls here: https://en.wikipedia.org/wiki/1901_college_football_season

In [58]:
# this will find the locations in the dataset where sponsor is null
bowlData.loc[bowlData['sponsor'].isna()]

id  year        date  day              winner_tie winner_rank  \
15      16  2006  12/30/2006  Sat                   Texas          18   
45      46  2005  12/23/2005  Fri                  Kansas         NaN   
48      49  2021  12/17/2021  Fri  Middle Tennessee State         NaN   
51      52  2017  12/22/2017  Fri                    Ohio         NaN   
59      60  2016  12/26/2016  Mon       Mississippi State         NaN   
...    ...   ...         ...  ...                     ...         ...   
1522  1523  2004  12/30/2004  Thu       Northern Illinois         NaN   
1523  1524  2003  12/30/2003  Tue            Fresno State         NaN   
1524  1525  2002  12/31/2002  Tue            Fresno State         NaN   
1525  1526  2001  12/31/2001  Mon          Michigan State         NaN   
1526  1527  2000  12/31/2000  Sun               Air Force         NaN   

      winner_points           loser_tie loser_rank  loser_points  attendance  \
15               26                Iowa        NaN            24     65875.0   
45               42             Houston        NaN            13     33505.0   
48               31              Toledo        NaN            24     13596.0   
51               41  Alabama-Birmingham        NaN             6     13585.0   
59               17               Miami        NaN            16     15717.0   
...             ...                 ...        ...           ...         ...   
1522             34                Troy        NaN            21     21456.0   
1523             17                UCLA        NaN             9     20126.0   
1524             30        Georgia Tech        NaN            21     10132.0   
1525             44        Fresno State         20            35     30456.0   
1526             37        Fresno State        NaN            34     26542.0   

                                                    mvp sponsor  \
15         QB Colt McCoy (Texas), DB Aaron Ross (Texas)     NaN   
45                            QB Jason Swanson (Kansas)     NaN   
48         MTSU QB Nicholas Vattiato, MTSU LB DQ Thomas     NaN   
51            Ohio RB Dorian Brown, Ohio FS Javon Hagan     NaN   
59                               MSU QB Nick Fitzgerald     NaN   
...                                                 ...     ...   
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...     NaN   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...     NaN   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...     NaN   
1525  WR Charles Rogers (Michigan State), DL Nick My...     NaN   
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...     NaN   

                                bowl_name  
15                             Alamo Bowl  
45                      Armed Forces Bowl  
48                           Bahamas Bowl  
51                           Bahamas Bowl  
59    Union Home Mortgage Gasparilla Bowl  
...                                   ...  
1522                  Silicon Valley Bowl  
1523                  Silicon Valley Bowl  
1524                  Silicon Valley Bowl  
1525                  Silicon Valley Bowl  
1526                  Silicon Valley Bowl  

[707 rows x 14 columns]

#### Rows Based on Column Comparison

If we want to return all the games where the winner was ranked below the loser, we could do it this way.

In [59]:
# finds all locations where the loser ranked below the winner
bowlData.loc[bowlData['loser_rank'] < bowlData['winner_rank']]

id  year        date  day      winner_tie winner_rank  winner_points  \
5        6  2016  12/29/2016  Thu  Oklahoma State          13             38   
7        8  2014    1/2/2015  Fri            UCLA          14             40   
13      14  2008  12/29/2008  Mon        Missouri          25             30   
21      22  2000  12/30/2000  Sat        Nebraska           9             66   
26      27  1995  12/28/1995  Thu       Texas A&M          19             22   
...    ...   ...         ...  ...             ...         ...            ...   
1422  1423  2001  12/28/2001  Fri           Texas           9             47   
1423  1424  2000  12/29/2000  Fri          Oregon           8             35   
1425  1426  1998  12/30/1998  Wed         Arizona           5             23   
1427  1428  1996  12/30/1996  Mon        Colorado           8             33   
1429  1430  1994  12/30/1994  Fri        Michigan          20             24   

           loser_tie loser_rank  loser_points  attendance  \
5           Colorado         11             8     59815.0   
7       Kansas State         11            35     60517.0   
13      Northwestern         22            23     55986.0   
21      Northwestern         18            17     60028.0   
26          Michigan         14            20     64597.0   
...              ...        ...           ...         ...   
1422      Washington         21            43     60548.0   
1423           Texas         12            30     63278.0   
1425        Nebraska         14            20     65354.0   
1427      Washington         13            21     54749.0   
1429  Colorado State         10            14     59453.0   

                                                    mvp  \
5        OSU WR James Washington, OSU DT Vincent Taylor   
7      RB Paul Perkins (UCLA), LB Eric Kendricks (UCLA)   
13    WR Jeremy Maclin (Missouri), LB Sean Witherspo...   
21    RB Dan Alexander (Nebraska), DL Kyle Vanden Bo...   
26    K Kyle Bryant (Texas A&M), LB Keith Mitchell (...   
...                                                 ...   
1422  QB Major Applewhite (Texas), LB Derrick Johnso...   
1423  QB Joey Harrington (Oregon), DB Rashad Bauman ...   
1425  QB Keith Smith (Arizona), DL Mike Rucker (Nebr...   
1427  QB Koy Detmer (Colorado), DL Nick Ziegler (Col...   
1429  QB Todd Collins (Michigan), QB Anthoney Hill (...   

                        sponsor     bowl_name  
5                        Valero    Alamo Bowl  
7                        Valero    Alamo Bowl  
13    Valero Energy Corporation    Alamo Bowl  
21                     Sylvania    Alamo Bowl  
26              Builders Square    Alamo Bowl  
...                         ...           ...  
1422                   Culligan  Holiday Bowl  
1423                   Culligan  Holiday Bowl  
1425                   Culligan  Holiday Bowl  
1427                   Plymouth  Holiday Bowl  
1429         Thrifty Car Rental  Holiday Bowl  

[255 rows x 14 columns]

Notice that when your output exceeds twenty lines Jupyter will format nicely and show a bit at the ending and a bit at the end. It will do the same thing if your output has too many columns.

### The df.query() Method:

We can also use a method called .query(). Inside the .query() we can put a query that looks a little bit like a database query.

#### Rows Where Column has Certain Numerical Value

In [60]:
query = "year == 1901"

# this will find the locations in the dataset where year is equal to 1901 using .query
bowlData.query(query)

id  year      date  day winner_tie winner_rank  winner_points  \
1020  1021  1901  1/1/1902  Wed   Michigan         NaN             49   

     loser_tie loser_rank  loser_points  attendance                      mvp  \
1020  Stanford        NaN             0      8000.0  RB Neil Snow (Michigan)   

     sponsor  bowl_name  
1020     NaN  Rose Bowl

#### Rows Where Column has String Value

In [61]:
# This will find the locations where the bowl_name is "Rose Bowl"
# notice the single quotes around the string "Rose Bowl"
bowlData.query("bowl_name == 'Rose Bowl'")

id  year      date  day           winner_tie winner_rank  \
913    914  2021  1/1/2022  Sat           Ohio State           7   
914    915  2020  1/1/2021  Fri              Alabama           1   
915    916  2019  1/1/2020  Wed               Oregon           7   
916    917  2018  1/1/2019  Tue           Ohio State           5   
917    918  2017  1/1/2018  Mon              Georgia           3   
...    ...   ...       ...  ...                  ...         ...   
1016  1017  1918  1/1/1919  Wed     Great Lakes Navy         NaN   
1017  1018  1917  1/1/1918  Tue  Mare Island Marines         NaN   
1018  1019  1916  1/1/1917  Mon               Oregon         NaN   
1019  1020  1915  1/1/1916  Sat     Washington State         NaN   
1020  1021  1901  1/1/1902  Wed             Michigan         NaN   

      winner_points            loser_tie loser_rank  loser_points  attendance  \
913              48                 Utah         10            45     87842.0   
914              31           Notre Dame          4            14     18373.0   
915              28            Wisconsin         11            27     90462.0   
916              28           Washington          9            23     91853.0   
917              54             Oklahoma          2            48     92844.0   
...             ...                  ...        ...           ...         ...   
1016             17  Mare Island Marines        NaN             0         NaN   
1017             19           Fort Lewis        NaN             7         NaN   
1018             14         Pennsylvania        NaN             0     27000.0   
1019             14                Brown        NaN             0     10000.0   
1020             49             Stanford        NaN             0      8000.0   

                                                    mvp              sponsor  \
913                    Ohio State WR Jaxon Smith-Njigba          Capital One   
914   Alabama WR DeVonta Smith, Alabama CB Patrick S...          Capital One   
915     Oregon QB Justin Herbert, Oregon S Brady Breeze  Northwestern Mutual   
916   Ohio State QB Dwayne Haskins, Ohio State S Bre...  Northwestern Mutual   
917     Georgia RB Sony Michel, Georgia LB Roquan Smith  Northwestern Mutual   
...                                                 ...                  ...   
1016                 WR George Halas (Great Lakes Navy)                  NaN   
1017         RB Hollis Huntington (Mare Island Marines)                  NaN   
1018                           OL John Beckett (Oregon)                  NaN   
1019                   RB Carl Dietz (Washington State)                  NaN   
1020                            RB Neil Snow (Michigan)                  NaN   

      bowl_name  
913   Rose Bowl  
914   Rose Bowl  
915   Rose Bowl  
916   Rose Bowl  
917   Rose Bowl  
...         ...  
1016  Rose Bowl  
1017  Rose Bowl  
1018  Rose Bowl  
1019  Rose Bowl  
1020  Rose Bowl  

[108 rows x 14 columns]

#### Rows Based on Substring Comparison

In [62]:
# Grab any row where the word "State" appears in the winner_tie column
bowlData.query('winner_tie.str.contains("State")', engine='python')

id  year        date  day        winner_tie winner_rank  \
3        4  2018  12/28/2018  Fri  Washington State          12   
5        6  2016  12/29/2016  Thu    Oklahoma State          13   
11      12  2010  12/29/2010  Wed    Oklahoma State          16   
14      15  2007  12/29/2007  Sat        Penn State         NaN   
17      18  2004  12/29/2004  Wed        Ohio State          24   
...    ...   ...         ...  ...               ...         ...   
1499  1500  2013  12/26/2013  Thu        Utah State         NaN   
1502  1503  2010  12/23/2010  Thu   San Diego State         NaN   
1523  1524  2003  12/30/2003  Tue      Fresno State         NaN   
1524  1525  2002  12/31/2002  Tue      Fresno State         NaN   
1525  1526  2001  12/31/2001  Mon    Michigan State         NaN   

      winner_points          loser_tie loser_rank  loser_points  attendance  \
3                28         Iowa State         25            26     60675.0   
5                38           Colorado         11             8     59815.0   
11               36            Arizona        NaN            10     57593.0   
14               24          Texas A&M        NaN            17     66166.0   
17               33     Oklahoma State        NaN             7     65265.0   
...             ...                ...        ...           ...         ...   
1499             21  Northern Illinois         24            14     23408.0   
1502             35               Navy        NaN            14     48049.0   
1523             17               UCLA        NaN             9     20126.0   
1524             30       Georgia Tech        NaN            21     10132.0   
1525             44       Fresno State         20            35     30456.0   

                                                    mvp  \
3     Washington State QB Gardner Minshew, Washingto...   
5        OSU WR James Washington, OSU DT Vincent Taylor   
11    WR Justin Blackmon (Oklahoma State), DB Markel...   
14    RB Rodney Kinlaw (Penn State), LB Sean Lee (Pe...   
17    WR Ted Ginn Jr. (Ohio State), DL Simon Fraser ...   
...                                                 ...   
1499  RB Joey DeMartino (Utah State), LB Jake Dought...   
1502  RB Ronnie Hillman (San Diego State), WR Vincen...   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...   
1525  WR Charles Rogers (Michigan State), DL Nick My...   

                            sponsor            bowl_name  
3                            Valero           Alamo Bowl  
5                            Valero           Alamo Bowl  
11        Valero Energy Corporation           Alamo Bowl  
14        Valero Energy Corporation           Alamo Bowl  
17                       MasterCard           Alamo Bowl  
...                             ...                  ...  
1499  San Diego County Credit Union      Poinsettia Bowl  
1502  San Diego County Credit Union      Poinsettia Bowl  
1523                            NaN  Silicon Valley Bowl  
1524                            NaN  Silicon Valley Bowl  
1525                            NaN  Silicon Valley Bowl  

[306 rows x 14 columns]

#### Rows Based on Column Comparison

In [63]:
# This will find all locations where the loser ranked below the winner using .query
bowlData.query("loser_rank < winner_rank")


id  year        date  day      winner_tie winner_rank  winner_points  \
5        6  2016  12/29/2016  Thu  Oklahoma State          13             38   
7        8  2014    1/2/2015  Fri            UCLA          14             40   
13      14  2008  12/29/2008  Mon        Missouri          25             30   
21      22  2000  12/30/2000  Sat        Nebraska           9             66   
26      27  1995  12/28/1995  Thu       Texas A&M          19             22   
...    ...   ...         ...  ...             ...         ...            ...   
1422  1423  2001  12/28/2001  Fri           Texas           9             47   
1423  1424  2000  12/29/2000  Fri          Oregon           8             35   
1425  1426  1998  12/30/1998  Wed         Arizona           5             23   
1427  1428  1996  12/30/1996  Mon        Colorado           8             33   
1429  1430  1994  12/30/1994  Fri        Michigan          20             24   

           loser_tie loser_rank  loser_points  attendance  \
5           Colorado         11             8     59815.0   
7       Kansas State         11            35     60517.0   
13      Northwestern         22            23     55986.0   
21      Northwestern         18            17     60028.0   
26          Michigan         14            20     64597.0   
...              ...        ...           ...         ...   
1422      Washington         21            43     60548.0   
1423           Texas         12            30     63278.0   
1425        Nebraska         14            20     65354.0   
1427      Washington         13            21     54749.0   
1429  Colorado State         10            14     59453.0   

                                                    mvp  \
5        OSU WR James Washington, OSU DT Vincent Taylor   
7      RB Paul Perkins (UCLA), LB Eric Kendricks (UCLA)   
13    WR Jeremy Maclin (Missouri), LB Sean Witherspo...   
21    RB Dan Alexander (Nebraska), DL Kyle Vanden Bo...   
26    K Kyle Bryant (Texas A&M), LB Keith Mitchell (...   
...                                                 ...   
1422  QB Major Applewhite (Texas), LB Derrick Johnso...   
1423  QB Joey Harrington (Oregon), DB Rashad Bauman ...   
1425  QB Keith Smith (Arizona), DL Mike Rucker (Nebr...   
1427  QB Koy Detmer (Colorado), DL Nick Ziegler (Col...   
1429  QB Todd Collins (Michigan), QB Anthoney Hill (...   

                        sponsor     bowl_name  
5                        Valero    Alamo Bowl  
7                        Valero    Alamo Bowl  
13    Valero Energy Corporation    Alamo Bowl  
21                     Sylvania    Alamo Bowl  
26              Builders Square    Alamo Bowl  
...                         ...           ...  
1422                   Culligan  Holiday Bowl  
1423                   Culligan  Holiday Bowl  
1425                   Culligan  Holiday Bowl  
1427                   Plymouth  Holiday Bowl  
1429         Thrifty Car Rental  Holiday Bowl  

[255 rows x 14 columns]

#### Rows Based on Comparison with a Variable

In [68]:
# First, let's get the mean winner_points x 2
twiceTheMean = bowlData.winner_points.mean() * 2

# then, we can use that value to query for winners with more than twice the mean
bigScorers = bowlData.query("winner_points > @twiceTheMean")


#### Combining Criteria

In [80]:
# Find all the rows where the winner is Alabama, 
# Alabama had more than 50 points, 
# and yet they weren't ranked number 1.
# TODO: Move this after the data corrections below

bowlData.query("(winner_tie == 'Alabama') and (winner_points > 50) and (winner_rank > 1)")

id  year      date  day winner_tie  winner_rank  winner_points  \
786  787  1952  1/1/1953  Thu    Alabama          8.0             61   

    loser_tie  loser_rank  loser_points  attendance  mvp sponsor    bowl_name  
786  Syracuse        16.0             6     66280.0  NaN     NaN  Orange Bowl

## Sorting Data

Let's sort by the winner's points to find the highest score by an upset winner.

The problem is that to sort properly, some of our rank data is not numeric as we saw above when we did .dtype.

Here's what happens if we sort on the winner_points column and try to sort descending (highest rank down to lowest):

In [72]:
# another cool thing - notice that we've asked for only certain columns by providing a list of the column
# names or 'series' that we want

bowlData[['id','year','winner_rank','mvp']].sort_values(by=['winner_rank'], ascending=False)

id  year   winner_rank  \
204    205  1949  Pennsylvania   
134    135  2019             9   
429    430  1984             9   
791    792  1947             9   
340    341  2017             9   
...    ...   ...           ...   
1522  1523  2004           NaN   
1523  1524  2003           NaN   
1524  1525  2002           NaN   
1525  1526  2001           NaN   
1526  1527  2000           NaN   

                                                    mvp  
204   RB Don Henigan (St. Vincent's (Pennsylvania)),...  
134                              Alabama WR Jerry Jeudy  
429   RB Thurman Thomas (Oklahoma State), QB Mike Ho...  
791                                                 NaN  
340   Penn State QB Trace McSorley, Penn State S Mar...  
...                                                 ...  
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...  
1523  RB Rodney Davis (Fresno State), DL Garrett McI...  
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...  
1525  WR Charles Rogers (Michigan State), DL Nick My...  
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...  

[1527 rows x 4 columns]

Notice that "Pennsylvania" is the highest winner_rank, but that's not a number at all. Let's fix that by changing that column to something numeric and then sorting again.

In [73]:
bowlData["winner_rank"] = pd.to_numeric(bowlData["winner_rank"])

ValueError: Unable to parse string "Pennsylvania" at position 204

Oh, we got an error. Let's fix that row to something numerical like zero.

In [74]:
# in the location where the row winner_rank is Pennsylvania and the column is winner_rank, set it 0
bowlData.loc[bowlData['winner_rank'] == 'Pennsylvania', ['winner_rank']] = 0

# then do the conversion
bowlData["winner_rank"] = pd.to_numeric(bowlData["winner_rank"])

Awesome, no error. Now let's convert the loser_rank column values to numeric:


In [75]:
bowlData["loser_rank"] = pd.to_numeric(bowlData["loser_rank"])

ValueError: Unable to parse string "TN" at position 207

Oops, the same error.  Please supply the code here to change "TN" to zero on the row where loser rank is zero.

In [76]:
# put the code here to change the value
bowlData.loc[bowlData['loser_rank'] == 'TN', ['loser_rank']] = 0

# then we can convert loser rank to numeric
bowlData["loser_rank"] = pd.to_numeric(bowlData["loser_rank"])

Now, we can see what we were looking for. We want to know who had the most points in a game when a lower ranking team beat a higher ranked team.

In [77]:
# get the rows where loser ranks lower than winner then sort by the winner's points descending
# notice how you can chain methods together
bowlData.query("loser_rank < winner_rank").sort_values(by=['winner_points'], ascending=False)

id  year        date  day       winner_tie  winner_rank  \
727    728  2011    1/4/2012  Wed    West Virginia         23.0   
920    921  2014    1/1/2015  Thu           Oregon          3.0   
917    918  2017    1/1/2018  Mon          Georgia          3.0   
1078  1079  1996    1/2/1997  Thu          Florida          3.0   
344    345  2013    1/1/2014  Wed  Central Florida         15.0   
...    ...   ...         ...  ...              ...          ...   
298    299  1939    1/1/1940  Mon          Clemson         15.0   
1153  1154  2008  12/31/2008  Wed     Oregon State         24.0   
456    457  1957  12/28/1957  Sat        Tennessee         18.0   
1133  1134  1941    1/1/1942  Thu          Fordham          8.0   
279    280  1958    1/1/1959  Thu        Air Force          8.0   

      winner_points        loser_tie  loser_rank  loser_points  attendance  \
727              70          Clemson        14.0            33     67563.0   
920              59    Florida State         2.0            20     91322.0   
917              54         Oklahoma         2.0            48     92844.0   
1078             52    Florida State         1.0            20     78344.0   
344              52           Baylor         6.0            42     65172.0   
...             ...              ...         ...           ...         ...   
298               6   Boston College        14.0             3     20000.0   
1153              3       Pittsburgh        18.0             0     49037.0   
456               3        Texas A&M         4.0             0     41160.0   
1133              2         Missouri         7.0             0     72000.0   
279               0  Texas Christian         7.0             0     75504.0   

                                                    mvp              sponsor  \
727                       QB Geno Smith (West Virginia)   Discover Financial   
920   QB Marcus Mariota (Oregon), LB Tony Washington...  Northwestern Mutual   
917     Georgia RB Sony Michel, Georgia LB Roquan Smith  Northwestern Mutual   
1078                        QB Danny Wuerffel (Florida)                Nokia   
344   QB Blake Bortles (UCF), LB Terrance Plummer (UCF)             Tostitos   
...                                                 ...                  ...   
298                          B Banks McFadden (Clemson)                  NaN   
1153             DL Victor Strong-Butler (Oregon State)                 Brut   
456   RB Bobby Gordon (Tennessee), RB John David Cro...                  NaN   
1133                                                NaN                  NaN   
279   OL Dave Phillips (Air Force), RB Jack Spikes (...                  NaN   

        bowl_name  
727   Orange Bowl  
920     Rose Bowl  
917     Rose Bowl  
1078   Sugar Bowl  
344   Fiesta Bowl  
...           ...  
298   Cotton Bowl  
1153     Sun Bowl  
456    Gator Bowl  
1133   Sugar Bowl  
279   Cotton Bowl  

[215 rows x 14 columns]

Nice, right? we can see that 23rd ranked West Virginia beat 14th ranked Clemson in 2011 with a score of 70!



## Grouping Data

In [84]:
avgPointsByWinners = bowlData.groupby("winner_tie")["winner_points"].mean()
avgPointsByWinners

winner_tie
Air Force             29.400000
Akron                 23.000000
Alabama               29.111111
Alabama-Birmingham    34.000000
Appalachian State     38.000000
                        ...    
Western Reserve       26.000000
William & Mary        20.000000
Wisconsin             27.055556
Wyoming               29.888889
Xavier                33.000000
Name: winner_points, Length: 166, dtype: float64

Notice that the list is sorted by winner_tie, the grouping column. This is the default behavior unless you specify sort=False as a parameter of the groupby() function. 

## Renaming Columns

To change column names, the easiest way is this:

In [85]:
# create a dictionary where key is the old name and value is the new name
columnMap = {"mvp":"Most Valuable Player", "attendance":"Attendance"}

bowlData = bowlData.rename(columns=columnMap, errors="raise")

bowlData

id  year        date  day         winner_tie  winner_rank  \
0        1  2021  12/29/2021  Wed           Oklahoma         14.0   
1        2  2020  12/29/2020  Tue              Texas         20.0   
2        3  2019  12/31/2019  Tue              Texas          NaN   
3        4  2018  12/28/2018  Fri   Washington State         12.0   
4        5  2017  12/28/2017  Thu    Texas Christian         13.0   
...    ...   ...         ...  ...                ...          ...   
1522  1523  2004  12/30/2004  Thu  Northern Illinois          NaN   
1523  1524  2003  12/30/2003  Tue       Fresno State          NaN   
1524  1525  2002  12/31/2002  Tue       Fresno State          NaN   
1525  1526  2001  12/31/2001  Mon     Michigan State          NaN   
1526  1527  2000  12/31/2000  Sun          Air Force          NaN   

      winner_points     loser_tie  loser_rank  loser_points  Attendance  \
0                47        Oregon        15.0            32     59121.0   
1                55      Colorado         NaN            23     10822.0   
2                38          Utah        12.0            10     60147.0   
3                28    Iowa State        25.0            26     60675.0   
4                39      Stanford        15.0            37     57653.0   
...             ...           ...         ...           ...         ...   
1522             34          Troy         NaN            21     21456.0   
1523             17          UCLA         NaN             9     20126.0   
1524             30  Georgia Tech         NaN            21     10132.0   
1525             44  Fresno State        20.0            35     30456.0   
1526             37  Fresno State         NaN            34     26542.0   

                                   Most Valuable Player sponsor  \
0     Oklahoma RB Kennedy Brooks, Oklahoma S Pat Fields  Valero   
1     Texas RB Bijan Robinson, Texas LB DeMarvion Ov...  Valero   
2          Texas QB Sam Ehlinger, Texas LB Joseph Ossai  Valero   
3     Washington State QB Gardner Minshew, Washingto...  Valero   
4               TCU QB Kenny Hill, TCU LB Travin Howard  Valero   
...                                                 ...     ...   
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...     NaN   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...     NaN   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...     NaN   
1525  WR Charles Rogers (Michigan State), DL Nick My...     NaN   
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...     NaN   

                bowl_name  
0              Alamo Bowl  
1              Alamo Bowl  
2              Alamo Bowl  
3              Alamo Bowl  
4              Alamo Bowl  
...                   ...  
1522  Silicon Valley Bowl  
1523  Silicon Valley Bowl  
1524  Silicon Valley Bowl  
1525  Silicon Valley Bowl  
1526  Silicon Valley Bowl  

[1527 rows x 14 columns]

## Handling Missing or Null Values
A common thing one needs to prepare for analysis or machine learning is to ensure that there are no null values. Let's see what's null in our bowl data. An easy thing to do is to get a number of null values per column like this:

In [86]:
# first, let's be sure we can see enough rows
pd.set_option('display.max_rows', 30)

# now show the count by column of null values
bowlData.isnull().sum()

id                        0
year                      0
date                      0
day                       0
winner_tie                0
winner_rank             773
winner_points             0
loser_tie                 0
loser_rank              856
loser_points              0
Attendance                9
Most Valuable Player    169
sponsor                 707
bowl_name                 0
dtype: int64

In [87]:
# this would return true or false for any row / column in which the result is missing / null
bowlData.isna()

id   year   date    day  winner_tie  winner_rank  winner_points  \
0     False  False  False  False       False        False          False   
1     False  False  False  False       False        False          False   
2     False  False  False  False       False         True          False   
3     False  False  False  False       False        False          False   
4     False  False  False  False       False        False          False   
...     ...    ...    ...    ...         ...          ...            ...   
1522  False  False  False  False       False         True          False   
1523  False  False  False  False       False         True          False   
1524  False  False  False  False       False         True          False   
1525  False  False  False  False       False         True          False   
1526  False  False  False  False       False         True          False   

      loser_tie  loser_rank  loser_points  Attendance  Most Valuable Player  \
0         False       False         False       False                 False   
1         False        True         False       False                 False   
2         False       False         False       False                 False   
3         False       False         False       False                 False   
4         False       False         False       False                 False   
...         ...         ...           ...         ...                   ...   
1522      False        True         False       False                 False   
1523      False        True         False       False                 False   
1524      False        True         False       False                 False   
1525      False       False         False       False                 False   
1526      False        True         False       False                 False   

      sponsor  bowl_name  
0       False      False  
1       False      False  
2       False      False  
3       False      False  
4       False      False  
...       ...        ...  
1522     True      False  
1523     True      False  
1524     True      False  
1525     True      False  
1526     True      False  

[1527 rows x 14 columns]

Notice the "True" in the winner_rank and loser_rank columns. This means some of the columns in "winner_rank" and "loser_rank" are missing.

Let's look at those rows with null values ("NaN")

In [88]:
# here we're asking for all the rows where winner_rank is null
bowlData[bowlData['winner_rank'].isnull()]

id  year        date  day         winner_tie  winner_rank  \
2        3  2019  12/31/2019  Tue              Texas          NaN   
9       10  2012  12/29/2012  Sat              Texas          NaN   
12      13  2009    1/2/2010  Sat         Texas Tech          NaN   
14      15  2007  12/29/2007  Sat         Penn State          NaN   
16      17  2005  12/28/2005  Wed           Nebraska          NaN   
...    ...   ...         ...  ...                ...          ...   
1522  1523  2004  12/30/2004  Thu  Northern Illinois          NaN   
1523  1524  2003  12/30/2003  Tue       Fresno State          NaN   
1524  1525  2002  12/31/2002  Tue       Fresno State          NaN   
1525  1526  2001  12/31/2001  Mon     Michigan State          NaN   
1526  1527  2000  12/31/2000  Sun          Air Force          NaN   

      winner_points       loser_tie  loser_rank  loser_points  Attendance  \
2                38            Utah        12.0            10     60147.0   
9                31    Oregon State        15.0            27     65277.0   
12               41  Michigan State         NaN            31     64757.0   
14               24       Texas A&M         NaN            17     66166.0   
16               32        Michigan        20.0            28     62000.0   
...             ...             ...         ...           ...         ...   
1522             34            Troy         NaN            21     21456.0   
1523             17            UCLA         NaN             9     20126.0   
1524             30    Georgia Tech         NaN            21     10132.0   
1525             44    Fresno State        20.0            35     30456.0   
1526             37    Fresno State         NaN            34     26542.0   

                                   Most Valuable Player  \
2          Texas QB Sam Ehlinger, Texas LB Joseph Ossai   
9     WR Marquise Goodwin (Texas), DL Alex Okafor (T...   
12    QB Taylor Potts (Texas Tech), DB Jamar Wall (T...   
14    RB Rodney Kinlaw (Penn State), LB Sean Lee (Pe...   
16     RB Cory Ross (Nebraska), DB Leon Hall (Michigan)   
...                                                 ...   
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...   
1525  WR Charles Rogers (Michigan State), DL Nick My...   
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...   

                        sponsor            bowl_name  
2                        Valero           Alamo Bowl  
9     Valero Energy Corporation           Alamo Bowl  
12    Valero Energy Corporation           Alamo Bowl  
14    Valero Energy Corporation           Alamo Bowl  
16                   MasterCard           Alamo Bowl  
...                         ...                  ...  
1522                        NaN  Silicon Valley Bowl  
1523                        NaN  Silicon Valley Bowl  
1524                        NaN  Silicon Valley Bowl  
1525                        NaN  Silicon Valley Bowl  
1526                        NaN  Silicon Valley Bowl  

[773 rows x 14 columns]

This method will replace anything that is NaN with zero:

In [89]:
# put data into a new data frame and replace all missing values in the winner_rank with zero
# notice that we are assigning the results of fillna to a new dataframe variable
betterBowlData = bowlData.fillna(0)

# let's see if anything is null - shouldn't be
betterBowlData.isna()

id   year   date    day  winner_tie  winner_rank  winner_points  \
0     False  False  False  False       False        False          False   
1     False  False  False  False       False        False          False   
2     False  False  False  False       False        False          False   
3     False  False  False  False       False        False          False   
4     False  False  False  False       False        False          False   
...     ...    ...    ...    ...         ...          ...            ...   
1522  False  False  False  False       False        False          False   
1523  False  False  False  False       False        False          False   
1524  False  False  False  False       False        False          False   
1525  False  False  False  False       False        False          False   
1526  False  False  False  False       False        False          False   

      loser_tie  loser_rank  loser_points  Attendance  Most Valuable Player  \
0         False       False         False       False                 False   
1         False       False         False       False                 False   
2         False       False         False       False                 False   
3         False       False         False       False                 False   
4         False       False         False       False                 False   
...         ...         ...           ...         ...                   ...   
1522      False       False         False       False                 False   
1523      False       False         False       False                 False   
1524      False       False         False       False                 False   
1525      False       False         False       False                 False   
1526      False       False         False       False                 False   

      sponsor  bowl_name  
0       False      False  
1       False      False  
2       False      False  
3       False      False  
4       False      False  
...       ...        ...  
1522    False      False  
1523    False      False  
1524    False      False  
1525    False      False  
1526    False      False  

[1527 rows x 14 columns]

## Writing Data to a File

To write data to a file, PANDAS has a built-in method to write .csv files - df.to_csv().

For example, let's write our bowl data to a file.

In [90]:
bowlData.to_csv(r'bowlData.csv', index = None, header=True)

## Learning more about Pandas

There is a great pandas "cheat sheet" in the lab files.